cluster-test/test-orchestrator/stat_ingestor.go

```diff
@@ -0,0 +1,223 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// The StatIngester is a UDP server that accepts ringpop stats with added
+// timestamps. The StatIngester analyzes the stream so that it knows when the
+// cluster reaches a stable state. It also writes the stream into a file for
+// later analysis.
+type StatIngester struct {
+	sync.Mutex
+	emptyNodes  map[string]bool
+	file        *os.File
+	wasUnstable bool
+}
+
+// NewStatIngester creates a new StatIngester
+func NewStatIngester() *StatIngester {
+	return &StatIngester{
+		emptyNodes: make(map[string]bool),
+	}
+}
+
+// WaitForStable blocks and waits until the cluster has reached a stable state.
+// waits for the cluster to first become unstable if it isn't already, and then
+// blocks until the cluster has reached a stable state again.
+func (si *StatIngester) WaitForStable(hosts []string) {
+	// wait for cluster to become unstable
+	for !si.wasUnstable {
+		time.Sleep(200 * time.Millisecond)
+	}
+	// wait for cluster to become stable
+	for !si.IsClusterStable(hosts) {
+		time.Sleep(200 * time.Millisecond)
+	}
+	si.wasUnstable = false
+}
+
+// IsClusterStable indicates, judging from the processed stats, whether the
+// cluster is in a stable state. The input are the hosts that should be
+// alive.
+func (si *StatIngester) IsClusterStable(hosts []string) bool {
+	si.Lock()
+	defer si.Unlock()
+
+	for _, h := range hosts {
+		hs := strings.Replace(h, ".", "_", -1)
+		hs = strings.Replace(hs, ":", "_", -1)
+		if empty, ok := si.emptyNodes[hs]; !ok || !empty {
+			return false
+		}
+	}
+	return true
+}
+
+// InsertLabel writes a label into the stats file.
+func (si *StatIngester) InsertLabel(label, cmd string) {
+	writeln(si.file, []byte(fmt.Sprintf("label:%s|cmd: %s", label, cmd)))
+}
+
+// statsQueue will receive all incomming stats for realtime-analysis.
+var statsQueue = make(chan []byte, 1024)
+
+// Listen starts listening on the specified port and writes the data into the
+// specified file.
+func (si *StatIngester) Listen(file string, port string) error {
+	// open output file
+	f, err := os.Create(file)
+	if err != nil {
+		return err
+	}
+	si.file = f
+
+	// start listening to stats
+	sAddr, err := net.ResolveUDPAddr("udp", ":"+port)
+	if err != nil {
+		return err
+	}
+
+	sConn, err := net.ListenUDP("udp", sAddr)
+	if err != nil {
+		return err
+	}
+
+	go si.startIngestion()
+
+	// handle stats
+	go func() {
+		buf := make([]byte, 1024)
+		for {
+			n, err := sConn.Read(buf)
+			if err != nil {
+				log.Fatalln(err)
+			}
+			if n == 0 {
+				return
+			}
+
+			// write to file
+			err = writeln(si.file, buf[0:n])
+			if err != nil {
+				log.Fatalln(err)
+			}
+
+			statsQueue <- []byte(string(buf[0:n]))
```

Comment:
Should we split the UDP code and the stats-parsing state machine to separate files/modules with a clear interface between the two?

I see two benefits:
1. Easier to add other listeners (e.g. stats replay?).
2. More modular, easier to understand code.